src/lmic/lmic.c

```diff
@@ -840,7 +840,7 @@ static bit_t decodeFrame (void) {
 
     // Process OPTS
     int m = LMIC.rssi - RSSI_OFF - getSensitivity(LMIC.rps);
-    LMIC.margin = m < 0 ? 0 : m > 254 ? 254 : m;
+    LMIC.margin = m < 0 ? 0 : m > 31 ? 31 : m;
```

Comment:
Thanks very much for finding this!  Unfortunately, I don't know how I can easily collaborate on your pull request other than to offer comments.  The change is a little bit more complicated.

1. the comment at line 841 needs to be *after* line 843 (please move to line 845), because this code is not logically related to the comment at line 841. It will clarify this if we move the comment now, as it will help reviewers.
2. According to the spec, the field in the DevStatusAns needs to be a 6-bit signed integer, and it's the SNR, not the margin (though it's named margin):

    > The margin (Margin) is the demodulation signal-to-noise ratio in dB rounded to the nearest integer value for the last successfully received DevStatusReq command. It is a signed integer of 6 bits with a minimum value of -32 and a maximum value of 31.

   Your change gets rid of the illegal value (good), but the DevStatusAns needs to transmit SNRs (different value, and negative much of thet time). Moreover, we need to capture the DevStatusAns SNR while decoding the DevStatusRequest.  We have the most recent snr in LMIC.snr, but it's not necessarily the correct one when we go to build the answer; so we need to stash it.

Thus, I suggest a more complicated fix.

- change code at line 843 as follows.

   ```c++
   // for legacy reasons, LMIC.margin is set to the unsigned sensitivity. It can never be negative.
   // it's only computed for legacy clients
   LMIC.margin = m < 0 ? 0 : m > 254 ? 254 : m;
   ```

- at line 610, change:
   ```c++
    case MCMD_DEVS_REQ: {
            LMIC.devsAns = 1;
            oidx += 1;
            continue;
   ```
   to
   ```c++
    case MCMD_DEVS_REQ: {
            LMIC.devsAns = 1;
            // LMIC.snr is SNR time 4, convert to real SNR; rounding towards zero.
            const int snr = (LMIC.snr + 2) / 4;
            // per [1.02] 5.5. the margin is the SNR. 
            LMIC.devsAnsMargin = (u1_t)(0x3F & (snr <= -32 ? -32 : snr >= 31 ? 31 : snr));
            oidx += 1;
            continue;
   ```

- Then, at line 1260, we need to change from `LMIC.frame[end+2] = LMIC.margin` to:

   ```c++
   // LMIC.frame[end+3] is set to the margin as used by the DevStatusAns 
   // (section 5.5 of 1.02b 5.3 of 1.1).
   // it is a 6-bit signed integer in -31..31, with the top two bits being RFU (zero).
   LMIC.frame[end+3] = LMIC.devAnsMargin;
   ```

- Finally, in `src/lmic/lmic.h` right after line 317 (where `devsAns` is defined), add:

   ```c++
   s1_t devAnsMargin;
   ```

Sorry, it's a lot! But this will make this more correct.

Reply:
Thank you for your writeup!
Your fix seems definitely like the best way to address the problem.

---

I have a couple of questions before I implement these changes:

1. Regarding the following lines,
    ``` 
    // LMIC.snr is SNR time 4, convert to real SNR; rounding towards zero.
    const int snr = (LMIC.snr + 2) / 4;
    ```
    Why is the `LMIC.snr` value incremented by 2? And should the `snr` variable be declared as `s2_t` rather than `int`?

2. Regarding the following change,
    ```
    LMIC.frame[end+3] = LMIC.devAnsMargin;
    ```
    Shouldn't the `LMIC.devAnsMargin` be in the 2nd byte of the `DevStatusAns` Payload rather than the 3rd?
    If not, then should the `end += 3;` at line 1267 be changed to `end += 4`?
    Maybe I'm just confused about how the `LMIC.frame` works..


3. In your comments you refer to ` [1.02]` or `[1.1]`, but I think the definition of margin has been the same [since v1.0](https://lora-alliance.org/resource-hub/lorawantm-specification-v10), so reporting the revision number is unnecessary.